### PR TITLE
TEP-0014 presents the Step timeout feature

### PR DESCRIPTION
**Pull request by Peaorl**
This PR introduces the `Step` timeout TEP.

An already open Pull Request related to this feature can be found [here](https://github.com/tektoncd/pipeline/pull/3087).

/cc @vdemeester @ImJasonH @sbwsg @bobcatfish

### CODE REVIEW

File: teps/0014-step-timeout.md
Comment (vdemeester) [inline teps/0014-step-timeout.md:42]: nit: should be `## Motivation`, no need for the `<a …`

File: teps/0014-step-timeout.md
Comment (vdemeester) [inline teps/0014-step-timeout.md:73]: nit: same here

**Review by sbwsg — APPROVED** [review]
I think the proposal looks good.  The one stumbling block I'm finding is that we have a lack of really concrete motivating use cases.  We've generally formulated the idea that one Step might need to do some setup and fail fast if it lingers, but it would be good to reify that with some more specific examples.  A couple that spring to mind:

- A dependency-fetching step can sometimes hang because the external registry slows down and it's better to fail and retry than wait for the connection to time out.
- A team has reduced the compilation time of their codebase and would like to ensure that new changes do not increase the compilation time substantially. They enforce this by setting a timeout on the compilation step in their build Task and run this Task against all new PRs.

**Comment by Peaorl** [comment]
Thanks @sbwsg! I added your examples to the TEP.

**Comment by ImJasonH** [comment]
This lgtm, I'll wait for someone else to give the final approval to have it merged.

Thanks for writing this up Alex!

File: teps/0014-step-timeout.md
Comment (popcor255) [inline teps/0014-step-timeout.md:27]: What does the spec look for the taskRun? I think the smallest change might be introducing this at the run level first.    If I use another person's task, Do I have to manually edit their task?

Comment (popcor255) [reply teps/0014-step-timeout.md:27]: This might be a dumb question, but this seems unclear to me.

Comment (ImJasonH) [reply teps/0014-step-timeout.md:27]: I think step-level details belong as a configuration of the Task, and should in general be opaque to TaskRun users. This lets a Task author change the implementation of their Task without requiring any change from users running that Task (hopefully 🤞).

If a Task author changes a step's timeout in a way that the TaskRun-er isn't comfortable with, they should fork and redefine the Task, or inline the TaskSpec definition in their TaskRun.

Comment (Peaorl) [reply teps/0014-step-timeout.md:27]: Indeed in that case you would have to edit the person's `Task`. The `TaskRun` spec doesn't change.

The reason for adding it to the `Task` and not the `TaskRun` is that it's assumed the `Task` author will have the most knowledge about how  the `Step` should behave.

Also latching on to [this](https://github.com/tektoncd/pipeline/pull/510#issuecomment-468378731) comment, it's suggested `TaskRuns` should be used for variations between invocations. I think in this case usually the timeout specification will remain the same between `TaskRun`/`PipelineRun` invocations.

Comment (popcor255) [reply teps/0014-step-timeout.md:27]: cool. That makes sense. It would also be pretty neat if as a task author, I could parameterize this field and give it a default value.

Comment (popcor255) [reply teps/0014-step-timeout.md:27]: I believe that substitution is done at run-state. So, if timeout field is typed for non-string. Im not sure if that would be possible. 🤔  (But, I might be getting to much in the weeds here) Anyway, this lgtm. 😄

Comment (Peaorl) [reply teps/0014-step-timeout.md:27]: Indeed that happens at run-time. 

In what scenario would you like to see this parametrized? Would that be to reuse a timeout specification across `Steps`.
Because I would say that every `Step` (that is desirable to timeout) should be scrutinized in order to find a fitting timeout for the `Step`. Therefore I'm not sure a timeout is something you would want to pass as a parameter in a `Task`.

Comment (popcor255) [reply teps/0014-step-timeout.md:27]: I have task that is a `bash script runner`. I provide it a git repo with a bash script and it runs it. I would like to decide how long the timeout is for that step. I don't want to use a task timeout because the first step is a git-clone. I would like to set a timeout on the script run step depending on what was pulled.

**Pull request merged**
